docs/Performance/promise.md

```diff
@@ -0,0 +1,195 @@
+
+* hand over the whole entity as object and construct the key from its data
+* provide an option to add additional data like filters to the call (mParameters as described in the [documentation](https://ui5.sap.com/#/api/sap.ui.model.odata.v2.ODataModel%23methods/read))
+* make sure its not possible to provide success / error callbacks with the parameters object as this would break the promise functionality
+
+The UI5 framework detects when the same OData call is done multiple times. It does only one request and returns the same result object for every call. If this result object is being changed, it changes for all calls. To prevent this from happening, the result is being copied with Object.assign()
+
+If you plan to use this with IE11, make sure to provide the [Object.assign Polyfill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#polyfill).
+
+```js
+/** 
+ * Returns the result of the OData call
+ * @param {object} payload - The payload which equals to the entity
+ * @param {object} parameters - The parameters like Filters added to the OData call
+ * @returns {Promise} Promise object represents result of the OData call
+ */
+function readMyEntity(payload, parameters) {
+  var model = this.getView().getModel();
+  parameters = parameters || {};
+  return model.metadataLoaded()
+    .then(function () {
+      return new Promise(function (resolve, reject) {
+        var key = model.createKey("/MyEntitySet", payload)
+        // prevent success / error callbacks to be overwritten
+        var params = Object.assign({}, parameters,
+          {
+            success: function (data, response) {
+              // prevent accidently change of response data for subsequent calls
+              var dataCopy = Object.assign({}, data);
+              resolve({ data: dataCopy, response: response });
+            },
+            error: function (error) {
+              // additional error handling when needed
+              reject(error);
+            }
+          });
+        model.read(key, params);
+        // update could look like this:
+        // model.update(key, payload, params);
+        // create could look like this:
+        // model.create("/MyEntitySet", payload, params);
+        // delete could look like this:
+        // model.remove(key, params);
+      })
+    })
+}
+
+readMyEntity({ id: 12345 })
+  .then(function (result) {
+    console.log(result.data, result.response);
+  })
+  .catch(function (error) {
+    console.error(error);
+  });
+```
+
+### Chaining multiple OData calls together
+
+Breaks after the first Promise which does not resolve.  
+The finally method is not supported by the Promise Polyfill for IE11.
+
+<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining>
+
+```js
+myElement.setBusy(true);
+readMyEntity({ id: 12345 })
+  .then(function (result) {
+    console.log(result.data, result.response);
+    return readMyEntity({ id: 67890 })
```

Comment:
probably worth explaining that explicitly returning an either async or sync value/function is key to stay in the promise chain

Reply:
maybe I did not got your point.
there is no such requirement to return a value to stay in the promise chain.

```js
Promise.resolve().then().then().then(() => console.log('demo'))
```

Reply:
because a non-specified `fn` is the same as an empty `return`
```js
Promise.resolve()
.then()
.then( () => {
    return
})
```
but:
```js
function someOtherFn() {
    // do stuff
    // NO RETURN!!!
}
Promise.resolve()
.then(someOtherFunc) // leaves the Promise chain intact aka implicit return
.then( () => {
    someOtherFunc() // breaks the Promise chain aka no return!
})
.then(() => true)
...
```

anyway: my point being is: always `return` sth in a `then`` in order to avoid creating unwanted side effects.

Reply:
I don't get it.
In your example, the promise chain resolves with true.

![image](https://user-images.githubusercontent.com/7511974/134897778-018dd680-84f9-43d9-af8b-fbdb20196948.png)

What unwanted side effects are you talking about?

I am a little bit confused right now... 🤔

Reply:
by "breaking the promise chain" i'm referring to unwanted side-effects.
hope this helps to illustrate my point:
```js
function someOtherFn(id = 2, timeout = 500) {
    return new Promise(resolve => setTimeout(() => {
        console.log(`${id}`)
        resolve()
    }), timeout)
}

Promise.resolve()
    .then(() => { console.log("0") }) // meeh :(
    // .then(() => { return console.log("0") }) // yeah :)
    .then(console.log("1")) // meeh :(
    // .then(() => { return console.log("1")}) // yeah :)
    .then(someOtherFn) 
    .then(() => {
        someOtherFn("3", 500) // meeh :( 
        // return someOtherFn("3", 0) // yeah :)
    })
    .then(() => {
        console.log("4")
    })

// meeh-state output:
// 1
// 0
// 2
// 4
// 3

// yeah-state output:
// 0
// 1
// 2
// 3
// 4
```

Reply:
resolved with a new commit